allow kubectl cmds to process files recursively, when given a dir

metral opened the pull request:
This PR accomplishes two things:
1. It creates a `--recursive` flag for use with certain `kubectl` commands that _currently_ do not process files beyond their first level of children, as seen in the issue https://github.com/kubernetes/kubernetes/issues/19767
2. It enables the ability to use the `--recursive` flag in the `kubectl` commands that currently _only_ support processing a directory up to its first level of children.

metral commented:
@bgrant0607 per your suggestion from https://github.com/kubernetes/kubernetes/issues/22342#issuecomment-196011686, I tried my luck on https://github.com/kubernetes/kubernetes/issues/19767

deads2k commented:
Squash, this will be all or nothing.

@smarterclayton This came up a year or so ago too.  Let me know how you feel about it and you can assign back to me for a review if you want.

deads2k commented on pkg/kubectl/resource/visitor.go:
```diff
@@ -396,19 +396,21 @@ func ExpandPathsToFileVisitors(mapper *Mapper, paths string, recursive bool, ext
 			if path != paths && !recursive {
 				return filepath.SkipDir
 			}
-			return nil
```
This return eliminated the need for the new conditionals you added, right?

Actually, are any of your changes here needed?

AdoHe replied:
@metral actually any code change here is not necessary. We have already consider multi level directory recursive.

metral replied:
@deads2k @AdoHe you are both correct. This section does not need to be edited - it was late last night when I thought that change was necessary. apologies for the confusion

bgrant0607 commented:
Cc @ghodds @kubernetes/kubectl

bgrant0607 commented:
Thanks, @metral!

@deads2k @smarterclayton Can you think of problems this would cause? It doesn't change default behavior. 

It seems like it would be convenient to facilitate organization of manifest files.

deads2k commented:
> @deads2k @smarterclayton Can you think of problems this would cause? It doesn't change default behavior.

I don't have a fundamental objection, but I'd like to see some sort of defense against cycles.  I'm also concerned about parity with things like `kubectl get` which take a `-f` and this would do weird things in those cases.

fabianofranz commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -331,6 +331,10 @@ func AddValidateFlags(cmd *cobra.Command) {
 }
 
+func AddRecursiveFlags(cmd *cobra.Command) {
+	cmd.Flags().Bool("recursive", false, "If true, process files recursively when given a directory")
```
I'd suggest a shorthand uppercase version of the flag, as supported in major unix commands like cp, ls, chmod, etc:

```
-R, --recursive
```

metral replied:
The shorthand `-R` has been added. Thanks for the feedback!

smarterclayton commented:
I don't have any objections, other than `-f` is less friendly for regular unix composition (with find and grep and others) than it should be, which is a separate issue.

deads2k commented:
Alright, so I'd like to see at least some defense against cycles and parity across every command that takes a `-f` for consistency.

metral commented:
> Alright, so I'd like to see at least some defense against cycles and parity across every command that takes a -f for consistency.

@deads2k any help with what you mean by defense cycles?

also, from what I can tell, some commands that take a `-f` need more work than what I've currently done with the easier, altered commands above. I wasn't sure if that would be in / out of scope of this PR, as I didn't want to introduce a big code change. Please advise as to how I should proceed.

deads2k commented:
> @deads2k any help with what you mean by defense cycles?

Right now I'm uncertain about behavior regarding hard or soft link cycles in a recursive directory traversal.  I'd like to see a unit test that proves that we don't get stuck in an infinite loop.



> some commands that take a -f need more work than what I've currently done with the easier, altered commands above. I wasn't sure if that would be in / out of scope of this PR, as I didn't want to introduce a big code change.

I'm against having an inconsistent set of flags for our commands.  It seems to me that if we're adding `-f` and we want to add `--recursive`, then we would always add them in a paired fashion and they should be respected by all commands.  From what you've shown, the change looks pretty mechanical.

metral commented:
> Right now I'm uncertain about behavior regarding hard or soft link cycles in a recursive directory traversal. I'd like to see a unit test that proves that we don't get stuck in an infinite loop.

A `Walk`, as seen in [ExpandPathsToFileVisitors](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource/visitor.go#L390), does not support soft/symbolic links per https://golang.org/src/path/filepath/path.go?s=11500:11545#L381

> some commands that take a -f need more work than what I've currently done with the easier, altered commands above. I wasn't sure if that would be in / out of scope of this PR, as I didn't want to introduce a big code change.

> I'm against having an inconsistent set of flags for our commands. It seems to me that if we're adding -f and we want to add --recursive, then we would always add them in a paired fashion and they should be respected by all commands. From what you've shown, the change looks pretty mechanical.

Yes, it is very mechanical change. However, issuing the same changes for commands `kubectl get` and `kubectl autoscale`, just to name a few, do not result in the same recursive functionality, using _only_ the minor changes I introduced in the likes of `kubectl create` and `kubectl stop`. So this is what I mean by needing further work. But I agree, `--recursive` should be respected by all commands - I'll have to think about the exceptions a bit longer, I suppose

metral commented:
I restored the visitor.go code to its original state - it was in fact already made recursively, so I removed my commits.

Still pending:

1. the rest of the `kubectl` cmds missing: (i.e. autoscale, patch, scale etc.)
2. appropriate unite tests

stay tuned

liggitt commented on pkg/kubectl/cmd/annotate.go:
```diff
@@ -114,6 +114,7 @@ func NewCmdAnnotate(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().StringVar(&options.resourceVersion, "resource-version", "", "If non-empty, the annotation update will only succeed if this is the current resource-version for the object. Only valid when specifying a single resource.")
 	usage := "Filename, directory, or URL to a file identifying the resource to update the annotation"
 	kubectl.AddJsonFilenameFlag(cmd, &options.filenames, usage)
+	cmdutil.AddRecursiveFlags(cmd)
```
Mismatch between how the flag is registered (encapsulated in a helper that assumes a global name) and used (literal string scattered in every command) is not a pattern we want to extend.

liggitt replied:
It would be better to have a "recursive" field in the options struct, bind the flag to that field (possibly with a helper to make the flag name centralized and consistent), and check the options struct field below

metral replied:
Thanks for the input @liggitt. I modeled the `--recursive` flag after the way that the Validate flag was [defined](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/helpers.go#L329) & is being [used](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create.go#L73).


I'll have to restructure the PR to use the model you've defined, so that I don't extend the pattern I misinterpreted.

I'm hoping to get to this sometime this week.

metral commented:
@deads2k here's a new PR that implements all of the commentators feedback, as well as, addresses the code style/pattern mismatch I had in my original PR

deads2k commented:
@k8s-bot test this please

ok to test

deads2k commented on pkg/kubectl/cmd/autoscale.go:
```diff
@@ -83,7 +83,7 @@ func RunAutoscale(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []
 		ContinueOnError().
 		NamespaceParam(namespace).DefaultNamespace().
-		FilenameParam(enforceNamespace, filenames...).
+		FilenameParam(enforceNamespace, false, filenames...).
```
Plumb something through to make this work.

deads2k commented on pkg/kubectl/cmd/expose.go:
```diff
@@ -117,7 +117,7 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 		ContinueOnError().
 		NamespaceParam(namespace).DefaultNamespace().
-		FilenameParam(enforceNamespace, options.Filenames...).
+		FilenameParam(enforceNamespace, false, options.Filenames...).
```
need to be wired up.

deads2k commented on pkg/kubectl/resource/builder_test.go:
```diff
@@ -234,32 +234,73 @@ func TestNodeBuilder(t *testing.T) {
 }
 
 func TestPathBuilderWithMultiple(t *testing.T) {
+	// This builder *should not* error
```
This test would be easier to read if it followed the 
```go
tests := []struct{
    name string
    directory string
    expectedNames []string
} {
    {"recursive-pod", "../.../../examples/recursive/pod", []string{"busybox-1", "busybox-2"}},
}

// iterate over tests, run each one separately, and verify results spearately
```

Right now this is a bit of an eye catcher.

deads2k commented on pkg/kubectl/resource/builder_test.go:
```diff
@@ -234,32 +234,73 @@ func TestNodeBuilder(t *testing.T) {
+	}
+
+	// This builder *should* error
+	b = NewBuilder(testapi.Default.RESTMapper(), api.Scheme, fakeClient(), testapi.Default.Codec()).
```
Separate test for these please.

Also, checking symlinks into the repo can cause problems for some tooling (people have yelled at me in other repos).  It sounds ugly, but constructing the borked structure inside your test, inside of a `/tmp` dir will keep the bad structures contained, but still allow us to run the test.  And don't forget to clean them up.

deads2k commented:
Pretty minor comments.  It's looking good.  I need to get up with @eparis to figure out the magic incantation for the bot.  I'd show him my previous comment, but I angered the mighty bot and it smashed my comment.

eparis commented:
ok to test

eparis commented:
Maybe the bot should record what nonsense you mere mortals utter in the log before it smashes you...

metral commented:
@deads2k i've gone ahead and wired up autoscale, expose & patch and cleaned up/refactored builder_test.go

kargakis commented on docs/man/man1/kubectl-stop.1:
```diff
@@ -46,6 +46,10 @@ If the resource is scalable it will be scaled to 0 before deletion.
 
 .PP
+\fB\-R\fP, \fB\-\-recursive\fP=false
+    If true, process directory recursively.
```
stop is deprecated, should we update it?

metral replied:
I'll defer to @deads2k

deads2k replied:
> stop is deprecated, should we update it?

May as well keep it consistent if it's not too painful.

kargakis commented on docs/man/man1/kubectl-expose.1:
```diff
@@ -92,6 +92,10 @@ labels are specified, the new service will re\-use the labels from the resource
 
 .PP
+\fB\-R\fP, \fB\-\-recursive\fP=false
+    If true, process directory recursively.
```
expose doesn't work with multiple resources currently, having the ability to recurse on a directory doesn't really make sense.

metral replied:
I modified expose to work with multiple resources, please see commits. Am I missing something?

deads2k commented on pkg/kubectl/cmd/expose.go:
```diff
@@ -147,101 +137,124 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
-		ports, err := f.PortsForObject(inputObject)
-		if err != nil {
-			return cmdutil.UsageError(cmd, fmt.Sprintf("couldn't find port via --port flag or introspection: %s", err))
+	for i := range infos {
```
I like this in principle, but I'd like to do it in a separate pull.  `--recursive` still makes sense for single resource commands, since I can point to a parent dir and this will navigate and find my single file at a lower level (its really the same argument as allowing directory references).

deads2k commented:
> I modified expose to work with multiple resources, please see commits. Am I missing something?

Do you mind pulling that work into a separate pull?  I like it in principle, but I'd like to do it in a separate pull. --recursive still makes sense for single resource commands, since I can point to a parent dir and this will navigate and find my single file at a lower level (its really the same argument as allowing directory references).

I think once its separated out, this will be good.

metral commented:
No problem. Should only the expose.go code be its own PR, or should all of the expose-related changes such as docs, and enabling the --recursive flag be in it as well?

Also, is this a good point to squash every thing else into a single commit?

deads2k commented:
> No problem. Should only the expose.go code be its own PR, or should all of the expose-related changes such as docs, and enabling the --recursive flag be in it as well?

Only the bits that switch us from single resource to multiple resources.  The recursive flag changes should all be here.

deads2k commented:
> Also, is this a good point to squash every thing else into a single commit?

Yeah, a squash is fine.

metral commented:
@deads2k: I've gone ahead and pulled out the 'multiple resources' capability from `expose.go` - I will submit its changes as a new PR after this 'recursive' PR is completed & merged. I also squashed & cleaned up this PR's commit msg

metral commented:
It seems that the e2e test is timing out due to what looks like some infrastructure endpoints not responding. I'm not sure if this is directly tied to my PR or not - please advise

deads2k commented on pkg/kubectl/cmd/autoscale.go:
```diff
@@ -108,62 +109,73 @@ func RunAutoscale(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []
-	params["scaleRef-kind"] = mapping.GroupVersionKind.Kind
-	params["scaleRef-name"] = name
-	params["scaleRef-apiVersion"] = mapping.GroupVersionKind.GroupVersion().String()
+	for i := range infos {
```
weren't you moving these changes into a separate pull?

metral replied:
apologies, i thought your comment about removing the multiple resources functionality was only limited to expose, not also autoscale & patch. I'll update it & push this change soon

deads2k commented on pkg/kubectl/cmd/patch.go:
```diff
@@ -127,31 +129,31 @@ func RunPatch(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []stri
-			// also, don't force the replacement.  If the replacement fails on a resourceVersion conflict, then it means this
-			// record hint is likely to be invalid anyway, so avoid the bad hint
-			resource.NewHelper(client, mapping).Replace(namespace, name, false, patchedObject)
+	for i := range infos {
```
And this one.

metral committed:
allow kubectl cmd to process dirs recursively

reqs:
    - the kubectl cmd must support the -f | --filename flag
    - the kubectl cmd must support visiting a dir one level deep,
    or using more than one resource

metral commented:
@deads2k 

> weren't you moving these changes into a separate pull?

this PR has been updated and is passing all tests. anything left?

deads2k commented:
lgtm

metral commented:
@deads2k any reason why the merge-robot continues to reissue tests?

deads2k commented:
> @deads2k any reason why the merge-robot continues to reissue tests?

The internal google tests (which block the merge queue) have had trouble today and your PR is retested before every attempt.  See http://submit-queue.k8s.io/
